### DISCUSSION TITLE
Motorized faders with limited range

### DISCUSSION THREAD

**Discussion by illuminancephoto** [rsjaffe/MIDI2LR#611] [Problems]
I'm trying my first controller with motorized faders (Behringer X-Touch Compact).  I like to use absolute value controls with a limited range.  For example by limiting the Exposure adjustment to +/- 3.00 in MIDI2LR Options, and assigning that to a controller that generates 0-60, with the MIDI2LR app configured to match, you get exposure varying from -3.00 to +3.00 in even 0.10 increments.  Nice.  As far as I can tell, though, when MIDI2LR sends the value back to the controller to position the fader, its interpreting the value as if the controller was using the full 0-127 range.  So my fader may be positioned at 30/127 rather than 30/60, which means the controller and the parameter value are always out of sync, and I end up needing pickup mode with my motorized faders.  Is there an easy way to make MIDI2LR send its position instruction scaled to a limited range?

**Comment by rsjaffe** [comment]
Yes, see https://github.com/rsjaffe/MIDI2LR/wiki/MIDI-Controller-Setup for how to set the controller range.

**Reply by illuminancephoto** [reply]
I understand how to limit the range in the app, and everything works as expected while adjusting the fader.  The hardware is configured to generate 0-60, and the MIDI2LR app is configured to match, absolute values, min 0 and max 60.  I get -3.00 to +3.00 with 0.10 increments when I move through the fader range, just like I want.  Its when MIDI2LR positions the motorized fader that things get out of sync.  Exposure of 0.00 should correspond to value 30 on my 0-60 limited control, and rather than being positioned half way as I'd like, it seems more like its being positioned at around 1/4 way through the range (-1.60 on my -3.00 to +3.00 scale).  I'm guessing its getting positioned at 30/127, causing the parameter value and the physical fader position to be out of sync.  If I set up the control to generate 35-95 instead of 0-60, and set the app to match, then I get the middle position to sync correctly, but the farther from 0.00 the parameter value is, the larger the sync error.

**Reply by rsjaffe** [reply]
Sounds like your fader is expecting 0-127 when setting a position. When you set the values to a limited range, MIDI2LR assumes that that is the range your controller uses. For example, if you set the CC control limits to 0-60, when MIDI2LR sees a 0 it assumes the controller is at 0% of its range, and when it sees a 60, it assumes the controller is at 100% of its range.

When data comes back from Lightroom, MIDI2LR _uses that same range_. So if a control in LR is at 0%, MIDI2LR will send a 0 to the MIDI controller, and when a control in LR is at 100%, MIDI2LR will send a 60 to the MIDI controller.

See https://github.com/rsjaffe/MIDI2LR/blob/020d8d822f9975b91132a7ff335ca0db12247099/src/application/ControlsModel.cpp#L260 for the details in how this is calculated.

**Reply by illuminancephoto** [reply]
Hmm.  Hadn't considered that changing the output range to 0-60 with X-Touch Editor doesn't also change the input range to move the fader.  That seems unfortunate but would definitely explain what's happening.  I suppose that's a dead end for me.  Use the full range or its never really going to work.  Thanks for your time though.  MIDI2LR is fantastic, and I appreciate your help!
